crates/aranya-daemon/src/test.rs

```diff
@@ -0,0 +1,415 @@
+//! Test module for aranya-daemon.
+
+#![allow(
+    clippy::expect_used,
+    clippy::indexing_slicing,
+    clippy::panic,
+    rust_2018_idioms
+)]
+
+use std::{
+    collections::BTreeMap,
+    fs,
+    net::Ipv4Addr,
+    ops::{Deref, DerefMut},
+    sync::Arc,
+};
+
+use anyhow::{bail, Context, Result};
+use aranya_crypto::{
+    default::{DefaultCipherSuite, DefaultEngine},
+    keystore::fs_keystore::Store,
+    Csprng, Rng,
+};
+use aranya_daemon_api::TeamId;
+use aranya_keygen::{KeyBundle, PublicKeys};
+use aranya_runtime::{
+    storage::linear::{libc::FileManager, LinearStorageProvider},
+    ClientState, GraphId,
+};
+use aranya_util::{ready, Addr};
+use s2n_quic::provider::tls::rustls::rustls::crypto::PresharedKey;
+use serial_test::serial;
+use tempfile::{tempdir, TempDir};
+use test_log::test;
+use tokio::{
+    sync::{
+        mpsc::{self, Receiver},
+        Mutex,
+    },
+    task::{self, AbortHandle},
+};
+
+use crate::{
+    actions::Actions,
+    aranya,
+    policy::{Effect, KeyBundle as DeviceKeyBundle, Role},
+    sync::{
+        self,
+        task::{quic::PskStore, PeerCacheKey, PeerCacheMap, SyncPeer},
+    },
+    vm_policy::{PolicyEngine, TEST_POLICY_1},
+    AranyaStore, InvalidGraphs,
+};
+
+// Aranya graph client for testing.
+pub type TestClient =
+    aranya::Client<PolicyEngine<DefaultEngine, Store>, LinearStorageProvider<FileManager>>;
+
+type TestState = sync::task::quic::State;
+// Aranya sync client for testing.
+pub type TestSyncer = sync::task::Syncer<TestState>;
+
+// Aranya sync server for testing.
+pub type TestServer = sync::task::quic::Server<
+    PolicyEngine<DefaultEngine, Store>,
+    LinearStorageProvider<FileManager>,
+>;
+
+pub struct TestDevice {
+    /// Aranya sync client.
+    pub syncer: TestSyncer,
+    /// The Aranya graph ID.
+    pub graph_id: GraphId,
+    /// The address that the server is listening on.
+    pub local_addr: Addr,
+    /// Aborts the server task.
+    handle: AbortHandle,
+    /// Public keys
+    pub pk: PublicKeys<DefaultCipherSuite>,
+    effect_recv: Receiver<(GraphId, Vec<Effect>)>,
+}
+
+impl TestDevice {
+    pub fn new(
+        client: TestClient,
+        server: TestServer,
+        local_addr: Addr,
+        pk: PublicKeys<DefaultCipherSuite>,
+        graph_id: GraphId,
+        psk_store: Arc<PskStore>,
+        caches: PeerCacheMap,
+    ) -> Result<Self> {
+        let waiter = ready::Waiter::new(1);
+        let notifier = waiter.notifier();
+        let handle = task::spawn(async { server.serve(notifier).await }).abort_handle();
```

Comment:
Should wait for the notification that the task has started:
```suggestion
        let handle = task::spawn(async { server.serve(notifier).await }).abort_handle();
        waiter.wait().await?;
```

Reply:
➕